libraries/AP_Baro/AP_Baro_KellerLD.cpp

```diff
@@ -61,37 +61,91 @@ AP_Baro_Backend *AP_Baro_KellerLD::probe(AP_Baro &baro, AP_HAL::OwnPtr<AP_HAL::D
         return false;
     }
+    hal.scheduler->delay(1);
 
-    _dev->get_semaphore()->take_blocking();
+    if(!_dev->transfer(nullptr, 0, recv, recvlen)) {
+        return false;
+    }
+    hal.scheduler->delay(1);
+    return true;
+}
 
-    // high retries for init
-    _dev->set_retries(10);
-    
-    bool cal_read_ok = true;
-    
+// This device has some undocumented finicky quirks and requires
+// delays when reading out the measurement range, but for some reason
+// this isn't an issue when requesting measurements.  This is why we
+// need to split the transfers with delays like this.  (Using
+// AP_HAL::I2CDevice::set_split_transfers will not work with these
+// sensors)
+bool AP_Baro_KellerLD::read_measurement_limit(float *limit, uint8_t msb_addr, uint8_t lsb_addr)
+{
     uint8_t data[3];
-    uint16_t ms_word, ls_word;
 
-    // This device has some undocumented finicky quirks and requires delays when reading out the
-    // measurement range, but for some reason this isn't an issue when requesting measurements.
-    // This is why we need to split the transfers with delays like this.
-    // (Using AP_HAL::I2CDevice::set_split_transfers will not work with these sensors)
-    
-    // Read out sensor P-mode and select relevant pressure offset
-    cal_read_ok &= _dev->transfer(&CMD_METADATA_PMODE, 1, nullptr, 0);
-    hal.scheduler->delay(1);
-    cal_read_ok &= _dev->transfer(nullptr, 0, &data[0], 3);
-    hal.scheduler->delay(1);
-    
+    if (!transfer_with_delays(&msb_addr, 1, &data[0], 3)) {
+        return false;
+    }
+
+    const uint16_t ms_word = (data[1] << 8) | data[2];
+    Debug("0x02x: %d [%d, %d, %d]", msb_addr, ms_word, data[0], data[1], data[2]);
+
+    if (!transfer_with_delays(&lsb_addr, 1, &data[0], 3)) {
+        return false;
+    }
+
+    const uint16_t ls_word = (data[1] << 8) | data[2];
+    Debug("0x%02x: %d [%d, %d, %d]", lsb_addr, ls_word, data[0], data[1], data[2]);
+
+    const uint32_t cal_data = (ms_word << 16) | ls_word;
+    memcpy(limit, &cal_data, sizeof(*limit));
+
+    if (isinf(*limit) || isnan(*limit)) {
+        return false;
+    }
+
+    Debug("data: %d, float: %.2f", cal_data, _p_min);
+    return true;
+}
+
+bool AP_Baro_KellerLD::read_cal()
+{
+    // Read out pressure measurement range
+    if (!read_measurement_limit(&_p_min, CMD_PRANGE_MIN_MSB, CMD_PRANGE_MIN_LSB)) {
+        return false;
+    }
+
+    if (!read_measurement_limit(&_p_max, CMD_PRANGE_MAX_MSB, CMD_PRANGE_MAX_LSB)) {
+        return false;
+    }
+
+    if (_p_max <= _p_min) {
+        return false;
+    }
+
+    return true;
+}
+
+// Read sensor P-Mode type and set pressure reference offset
+// This determines the pressure offset based on the type of sensor
+// vented to atmosphere, gauged to vacuum, or gauged to standard sea-level pressure
+bool AP_Baro_KellerLD::read_mode_type()
+{
+    uint8_t cmd = CMD_METADATA_PMODE;
+    uint8_t data[3];
+    if (!transfer_with_delays(&cmd, 1, &data[0], 3)) {
```

Comment:
Pre-existing, but should probably be `ARRAY_SIZE(data)`.  And just pass in data....

Reply:
Done